File: Leetcode1895. Largest Magic Square/Solution.py
class Solution:
    def largestMagicSquare(self, grid):
        total_rows, total_cols = len(grid), len(grid[0])

        # Prefix sum arrays
        row_prefix_sum = [[0] * (total_cols + 1) for _ in range(total_rows)]
        col_prefix_sum = [[0] * total_cols for _ in range(total_rows + 1)]
        main_diag_prefix = [[0] * (total_cols + 1) for _ in range(total_rows + 1)]
        anti_diag_prefix = [[0] * (total_cols + 2) for _ in range(total_rows + 1)]

        # Build prefix sums
        for row in range(total_rows):
            for col in range(total_cols):
                row_prefix_sum[row][col + 1] = row_prefix_sum[row][col] + grid[row][col]
                col_prefix_sum[row + 1][col] = col_prefix_sum[row][col] + grid[row][col]
                main_diag_prefix[row + 1][col + 1] = main_diag_prefix[row][col] + grid[row][col]
                anti_diag_prefix[row + 1][col] = anti_diag_prefix[row][col + 1] + grid[row][col]

        def is_magic_square(start_row, start_col, size):
            target_sum = row_prefix_sum[start_row][start_col + size] - row_prefix_sum[start_row][start_col]

            # Check all rows
            for r in range(start_row, start_row + size):
                if row_prefix_sum[r][start_col + size] - row_prefix_sum[r][start_col] != target_sum:
                    return False

            # Check all columns
            for c in range(start_col, start_col + size):
                if col_prefix_sum[start_row + size][c] - col_prefix_sum[start_row][c] != target_sum:
                    return False

            # Check both diagonals
            if main_diag_prefix[start_row + size][start_col + size] - main_diag_prefix[start_row][start_col] != target_sum:
                return False

            if anti_diag_prefix[start_row + size][start_col] - anti_diag_prefix[start_row][start_col + size] != target_sum:
                return False

            return True

        max_possible_size = min(total_rows, total_cols)

        # Try larger squares first
        for size in range(max_possible_size, 0, -1):
            for row in range(total_rows - size + 1):
                for col in range(total_cols - size + 1):
                    if is_magic_square(row, col, size):
                        return size

        return 1

File: Leetcode1895. Largest Magic Square/Solution.cpp
class Solution {
public:
    int largestMagicSquare(vector<vector<int>>& grid) {
        int totalRows = grid.size();
        int totalCols = grid[0].size();

        // Prefix sum arrays
        vector<vector<int>> rowPrefixSum(totalRows, vector<int>(totalCols + 1, 0));
        vector<vector<int>> colPrefixSum(totalRows + 1, vector<int>(totalCols, 0));
        vector<vector<int>> mainDiagPrefix(totalRows + 1, vector<int>(totalCols + 1, 0));
        vector<vector<int>> antiDiagPrefix(totalRows + 1, vector<int>(totalCols + 2, 0));

        // Build prefix sums
        for (int row = 0; row < totalRows; row++) {
            for (int col = 0; col < totalCols; col++) {
                rowPrefixSum[row][col + 1] =
                    rowPrefixSum[row][col] + grid[row][col];

                colPrefixSum[row + 1][col] =
                    colPrefixSum[row][col] + grid[row][col];

                mainDiagPrefix[row + 1][col + 1] =
                    mainDiagPrefix[row][col] + grid[row][col];

                antiDiagPrefix[row + 1][col] =
                    antiDiagPrefix[row][col + 1] + grid[row][col];
            }
        }

        int maxPossibleSize = min(totalRows, totalCols);

        // Try larger squares first
        for (int size = maxPossibleSize; size >= 1; size--) {
            for (int row = 0; row + size <= totalRows; row++) {
                for (int col = 0; col + size <= totalCols; col++) {
                    if (isMagicSquare(row, col, size,
                                      rowPrefixSum, colPrefixSum,
                                      mainDiagPrefix, antiDiagPrefix)) {
                        return size;
                    }
                }
            }
        }

        return 1;
    }

private:
    bool isMagicSquare(
        int startRow,
        int startCol,
        int size,
        vector<vector<int>>& rowPrefixSum,
        vector<vector<int>>& colPrefixSum,
        vector<vector<int>>& mainDiagPrefix,
        vector<vector<int>>& antiDiagPrefix) {

        int targetSum =
            rowPrefixSum[startRow][startCol + size]
          - rowPrefixSum[startRow][startCol];

        // Check all rows
        for (int r = startRow; r < startRow + size; r++) {
            int rowSum =
                rowPrefixSum[r][startCol + size]
              - rowPrefixSum[r][startCol];
            if (rowSum != targetSum) {
                return false;
            }
        }

        // Check all columns
        for (int c = startCol; c < startCol + size; c++) {
            int colSum =
                colPrefixSum[startRow + size][c]
              - colPrefixSum[startRow][c];
            if (colSum != targetSum) {
                return false;
            }
        }

        // Check main diagonal
        int mainDiagSum =
            mainDiagPrefix[startRow + size][startCol + size]
          - mainDiagPrefix[startRow][startCol];
        if (mainDiagSum != targetSum) {
            return false;
        }

        // Check anti-diagonal
        int antiDiagSum =
            antiDiagPrefix[startRow + size][startCol]
          - antiDiagPrefix[startRow][startCol + size];
        if (antiDiagSum != targetSum) {
            return false;
        }

        return true;
    }
};
